homeassistant/components/bluetooth/passive_update_processor.py

```diff
@@ -62,6 +150,111 @@ def update(self, new_data: PassiveBluetoothDataUpdate[_T]) -> None:
+            "devices": {
+                key or "": device_info for key, device_info in self.devices.items()
+            },
+            "entity_descriptions": {
+                key.to_string(): serialize_entity_description(description)
+                for key, description in self.entity_descriptions.items()
+            },
+            "entity_names": {
+                key.to_string(): name for key, name in self.entity_names.items()
+            },
+            "entity_data": {
+                key.to_string(): data for key, data in self.entity_data.items()
+            },
+        }
+
+    @callback
+    def async_set_restore_data(
+        self,
+        restore_data: RestoredPassiveBluetoothDataUpdate,
+        entity_description_class: type[EntityDescription],
+    ) -> None:
+        """Set the restored data from storage."""
+        self.devices.update(
+            {
+                key or None: device_info
+                for key, device_info in restore_data["devices"].items()
+            }
+        )
+        self.entity_descriptions.update(
+            {
+                PassiveBluetoothEntityKey.from_string(
+                    key
+                ): deserialize_entity_description(entity_description_class, description)
+                for key, description in restore_data["entity_descriptions"].items()
+                if description
+            }
+        )
+        self.entity_names.update(
+            {
+                PassiveBluetoothEntityKey.from_string(key): name
+                for key, name in restore_data["entity_names"].items()
+            }
+        )
+        self.entity_data.update(
+            {
+                PassiveBluetoothEntityKey.from_string(key): cast(_T, data)
+                for key, data in restore_data["entity_data"].items()
+            }
+        )
+
+
+def async_register_coordinator_for_restore(
+    hass: HomeAssistant, coordinator: PassiveBluetoothProcessorCoordinator
+) -> CALLBACK_TYPE:
+    """Register a coordinator to have its processors data restored."""
+    data: PassiveBluetoothProcessorData = hass.data[PASSIVE_UPDATE_PROCESSOR]
+    data.coordinators.add(coordinator)
+    if restore_key := coordinator.restore_key:
+        coordinator.restore_data = data.all_restore_data.setdefault(restore_key, {})
+
+    @callback
+    def _unregister_coordinator_for_restore() -> None:
+        """Unregister a coordinator."""
+        data.coordinators.remove(coordinator)
+
+    return _unregister_coordinator_for_restore
+
+
+async def async_setup(hass: HomeAssistant) -> None:
+    """Set up the passive update processor coordinators."""
+    data = PassiveBluetoothProcessorData()
+    hass.data[PASSIVE_UPDATE_PROCESSOR] = data
+    storage: Store[dict[str, dict[str, RestoredPassiveBluetoothDataUpdate]]] = Store(
+        hass, 1, STORAGE_KEY
+    )
+    if restore_data := await storage.async_load():
+        data.all_restore_data = restore_data
+    coordinators = data.coordinators
```

Comment:
Would be cleaner to create the values rather than rely on the factory behavior so we know who owns the source